### ISSUE TITLE
Add debugWebViewPlatformProvider for WebViewPlatform

### ISSUE THREAD

**Issue by SimonWeidemann** [flutter/flutter#128971]
### Is there an existing issue for this?

- [X] I have searched the [existing issues](https://github.com/flutter/flutter/issues)
- [X] I have read the [guide to filing a bug](https://flutter.dev/docs/resources/bug-reports)

### Use case

If we set the WebViewPlatform instance in one test, we cannot set it back to null after the instance is no longer needed.
```
  /// Platform-specific plugins should set this with their own platform-specific
  /// class that extends [WebViewPlatform] when they register themselves.
  static set instance(WebViewPlatform? instance) {
    if (instance == null) {
      throw AssertionError(
          'Platform interfaces can only be set to a non-null instance');
    }
    PlatformInterface.verify(instance, _token);
    _instance = instance;
  }
```

So if we mock the WebViewPlatform in one test, the other test would deal with the side effect of setting the mock.

```
class MockWebViewPlatform extends Mock
    with MockPlatformInterfaceMixin
    implements WebViewPlatform {}

main(){
  group('web view platforms do not have side effects', () {
    late final bool testWebViewPlatformWasCalled;

    test('Set instance in first test', () {
      testWebViewPlatformWasCalled = true;
      WebViewPlatform.instance = MockWebViewPlatform();
    });

    test('check if the WebViewPlatform instance is null', () {
      expect(
        testWebViewPlatformWasCalled,
        true,
        reason: 'We have to verify that the testWebViewPlatformWasCalled'
            ' is once executed so we can see if this affect this test',
      );
      expect(WebViewPlatform.instance, null);
    });
  });
}
```
The result of the second test would be as follows.
```
This test Fails with: 
Expected: <null>
Actual: MockWebViewPlatform:<MockWebViewPlatform>
```

This may cause undesired behaviour because the mock is still present. 

One solution would be to simply generate a new mock and set it, but we could not go back to the 'initial' state. 

### Proposal

 I would suggest something like the `debugNetworkImageHttpClientProvider`.
 
 flutter/packages/flutter/lib/src/painting/debug.dart

Somewhere there is a debugWebViewPlatformProvider field:
```
WebViewPlatform Function()? debugWebViewPlatformProvider;
```
And when we try to get the instance, we use it like this:
```
  /// The instance of [WebViewPlatform] to use.
  static WebViewPlatform? get instance {
    WebViewPlatform? result = _instance;
    assert(() {
      if (debugWebViewPlatformProvider != null) {
        final WebViewPlatform platform = debugWebViewPlatformProvider!();
        PlatformInterface.verify(platform, _token);
        result = platform;
      }
      return true;
    }());
    return result;
  }
```

**Comment by stuartmorgan** [comment]
It looks like you are trying to prevent test pollution between tests; normally we handle this in our own tests by setting a fake/mock/dummy implementation fresh for every test in `setUp`. Is there a reason that doesn't work for your use case?

If so, could you elaborate on the expected behavior of the "debug" implementation? Would it provide a benefit over allowing setting back to null?

**Comment by SimonWeidemann** [comment]
Consider the following, you have a test file with several groups. In the group at the beginning of your file, you set the MockWebViewPlatform instance in the group's setUp method.

A few groups later, you interact with the instance again, unaware that the instance has been set. When you check the setUp of the group and the inherited setUps, it is not visible that the instance has been set.

So by default, I would say it would be good practice to always set the instance that survives multiple tests and is not a globally visible field in the testfile to a non-functional instance or null in the tearDwon method. That way, accessing it again will at least throw an exception, and it will be clear that you are accessing something that needs a proper test setup.

As mentioned above, a non-functional instance can also serve the purpose of giving the tester feedback that an exchange has taken place. However, the error message is not as informative as the default one that is thrown if you access it with null. Otherwise, you have to figure out from the stack trace that a substitution has been made, and if you want to interact with it, you have to replace/use the current one.

**Comment by stuartmorgan** [comment]
> So by default, I would say it would be good practice to always set the instance that survives multiple tests and is not a globally visible field in the testfile to a non-functional instance or null in the tearDwon method.

Fair enough; we could probably change this to allow setting back to `null` (either by allowing null here, or to avoid accidental usage of `null` in production a `@visibleForTesting` method to reset it).

As a workaround for now, you can add something like:
```
class TombstoneWebViewPlatform extends WebViewPlatform {}
```

to your test file, and set `instance` to that in your teardown.

> However, the error message is not as informative as the default one that is thrown if you access it with null.

It sounds like you would prefer being able to set it back to `null` rather than having a special built-in "debug" instance then? If so I can retitle this issue.

**Comment by SimonWeidemann** [comment]
If you can give us a solution that allows us to set it to null. I would prefer that.

I chose the debugWebViewPlatformProvider interceptor because the assertion guarantees that it is not available in prod.

`@visibleForTesting` is just an annotation for the analyser, which can still be bypassed with and : `// ignore: invalid_use_of_visible_for_testing_member`. 

But I leave the way to achieve this up to you.

**Comment by stuartmorgan** [comment]
> I chose the debugWebViewPlatformProvider interceptor because the assertion guarantees that it is not available in prod.
> 
> `@visibleForTesting` is just an annotation for the analyser, which can still be bypassed with and : `// ignore: invalid_use_of_visible_for_testing_member`.

If someone wants to go out of their way to break webview in their own webview-using application in production, they can already trivially do so (such as via the line in my comment just above), so I'm not sure what the concern is here.
